docs/03-Implementation-Guide/01-Payload-Processing.md

```diff
@@ -20,6 +20,20 @@ encouraged to use [Base64url](https://tools.ietf.org/html/rfc4648#page-7)
 encoding to preserve the original intended binary data, but still use UTF-8 for
 the actual payloads.
 
+## Type Safety with Cryptographic Keys
+
+PASETO library implementations **MUST** implement some means of preventing type
+confusion bugs between different cryptography keys. For example:
+
+* Prepending each key in memory with a magic byte to serve as a type indicator
+  (distinct for every combination of version and purpose).
+* In object-oriented programming languages, using separate classes for each
+  cryptography key object that may share an interface or common base class.
+
+It **MUST NOT** be possible for a user to take a known public key (used by
```

Comment:
Being pedantic here: more generic rules are given in the example than in the paragraph following it. E.g. you probably also don't want the confusion to occur the other way round either i.e. *local* keys shouldn't be used with *public* parts of the API (only the converse direction is stated :) ) – perhaps obviously symmetric language would be beneficial here: e.g. "*local* purpose keys must not be confused with *public* purpose keys". Similarly nor should keys be able to be confused across different protocol versions.

Maybe something like or along the lines of:
```markdown
Cryptographic keys **MUST** require the user to state a version and a purpose for which they
will be used. Furthermore, given a cryptographic key it **MUST NOT** be possible for a user
to use this key for any version and purpose combination other than that which was specified
during the creation of this key.
```

Also, by "public key" I guess you mean the program's representation of one (which would allow for specification of intent), but not its raw material alone (otherwise all raw material would require unique lengths?)

Reply:
Yes, absolutely. I appreciate your pedantry here :)

Reply:
I think you forgot to update this file in 0e53c72 :)